### PR TITLE
[IMP] sale_timesheet: Add test to check analytic_account_balance on project

### PR DESCRIPTION
**Pull request by RaoufGhrissi**
This commit adds a unit test to check if the value of 
`analytic_account_balance` field in a project is the one expected when a 
timesheet is added into this project.
The analytic account balance value expected should be equal to 
`timesheet_cost * timesheet_unit_amount * -1`, where:
- `timesheet_cost`: the employee cost or the cost set on an employee mapping 
                    defined in the project.
- `timesheet_unit_amount`: the number of hours/days set on the timesheet.

task-2783686

### CODE REVIEW

File: addons/sale_timesheet/tests/test_project.py [hunk 1]
```diff
@@ -97,3 +97,52 @@ def test_fetch_sale_order_items(self):
...
+
+    def test_analytic_account_balance(self):
+        """
+            1) Add new billable project
+            2) Add Employee/SOL mapping in the project
+            3) Add Task and Timesheet with the same user
+            4) Assert analytic_account_balance is calculated
+        """
+        partner = self.env['res.partner'].search([('name', '=', 'Deco Addict')], limit=1)
+        employee = self.env['hr.employee'].search([], limit=1)
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_project.py:109]: You should use the data created in the test instead of using the demo data. 🙂

File: addons/sale_timesheet/tests/test_project.py [hunk 2]
```diff
@@ -97,3 +97,52 @@ def test_fetch_sale_order_items(self):
...
+        self.assertTrue(partner)
+        self.assertTrue(employee)
+
+        project = self.env['project.project'].create({
+            'name': 'Project A',
+            'allow_billable': True,
+            'partner_id': partner.id
+        })
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_project.py:117]: You should use the project create in a common class instead of a creating a new one.

File: addons/sale_timesheet/tests/test_project.py [hunk 3]
```diff
@@ -97,3 +97,52 @@ def test_fetch_sale_order_items(self):
...
+        sale_line = self.env['sale.order.line'].search([('order_partner_id', '=', partner.id),
+                ('name', '=', 'Junior Architect (Invoice on Timesheets)')], limit=1)
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_project.py:119]: You should use the SOL defined in the common class instead of searching a demo data.

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I have made some comments.

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I have made some comments.

File: addons/sale_project/tests/test_sale_project.py [hunk 4]
```diff
@@ -3,7 +3,6 @@
 
 from odoo.tests.common import TransactionCase
 
-
```
Comment (xavierbol) [inline addons/sale_project/tests/test_sale_project.py:6]: Unnecessary diff

File: addons/sale_timesheet/tests/test_project.py [hunk 5]
```diff
@@ -97,3 +97,55 @@ def test_fetch_sale_order_items(self):
...
+    def test_analytic_account_balance(self):
+        """
+            1) Add new billable project
+            2) Add Employee/SOL mapping in the project
+            3) Add Task and Timesheet with the same user
+            4) Assert analytic_account_balance is calculated
+        """
+        partner = self.env['res.partner'].create({
+            'name': 'Partner',
+        })
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_project.py:110]: You could use a partner created in the common class that we inherit, I think we have those partner: <https://github.com/odoo/odoo/blob/f34d8632572dfe8911b62471882eae60f6e6ae72/addons/account/tests/common.py#L143-L159>

File: addons/sale_timesheet/tests/test_project.py [hunk 6]
```diff
@@ -97,3 +97,55 @@ def test_fetch_sale_order_items(self):
...
+        unit_cost = 15
+        unit_amount = 6
+        expected_analytic_account_balance = unit_cost * unit_amount * -1
+        self.project_global.write({
+            'sale_line_id': sale_line.id,
+            'sale_line_employee_ids': [
+                Command.create({
+                    'employee_id': self.employee_user.id,
+                    'sale_line_id': sale_line.id,
+                    'cost': unit_cost
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_project.py:133]:
```suggestion
                    'cost': unit_cost,
```

File: addons/sale_timesheet/tests/test_project.py [hunk 7]
```diff
@@ -97,3 +97,55 @@ def test_fetch_sale_order_items(self):
...
+            })]
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_project.py:134]:
```suggestion
            })],
```

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I have made some comments and after that, it will look good to me. 🙂

File: addons/sale_timesheet/tests/test_project.py [hunk 8]
```diff
@@ -129,3 +129,53 @@ def test_compute_cost_in_employee_mappings(self):
...
+
+    def test_analytic_account_balance(self):
+        """
+            1) Add new billable project
+            2) Add Employee/SOL mapping in the project
+            3) Add Task and Timesheet with the same user
+            4) Assert analytic_account_balance is calculated
+        """
+
+        sale_order = self.env['sale.order'].create({
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_project.py:141]: No need to add an additional blank line here. 🙂

File: addons/sale_timesheet/tests/test_project.py [hunk 9]
```diff
@@ -129,3 +129,53 @@ def test_compute_cost_in_employee_mappings(self):
...
+        unit_cost = 15
+        unit_amount = 6
+        expected_analytic_account_balance = unit_cost * unit_amount * -1
+        self.project_global.write({
+            'sale_line_id': sale_line.id,
+            'sale_line_employee_ids': [
+                Command.create({
+                    'employee_id': self.employee_user.id,
+                    'sale_line_id': sale_line.id,
+                    'cost': unit_cost,
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_project.py:163]: Since the `timesheet_cost` of the employee is also equal to 15, then we could remove this line since the cost will automatically equals to the timesheet cost of the employee linked to this mapping. 🤔 
What do you think? 🙂

File: addons/sale_timesheet/tests/test_project.py [hunk 10]
```diff
@@ -129,3 +129,53 @@ def test_compute_cost_in_employee_mappings(self):
...
+            'product_id': self.product_delivery_timesheet1.id,
+            'product_uom_qty': 1,
+            'product_uom': self.product_delivery_timesheet1.uom_id.id,
+            'price_unit': self.product_delivery_timesheet1.list_price,
+            'order_id': sale_order.id,
+        })
+
+        unit_cost = 15
+        unit_amount = 6
+        expected_analytic_account_balance = unit_cost * unit_amount * -1
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_project.py:156]: Since the timesheet cost of the employee used in the creation of a employee mapping for the project below is also equal to 15. We could rewrite like this:
```suggestion
        unit_amount = 6
        expected_analytic_account_balance = self.employee_user.timesheet_cost * unit_amount * -1
```

What do you think? 🙂

File: addons/sale_timesheet/tests/test_project.py [hunk 11]
```diff
@@ -129,3 +129,53 @@ def test_compute_cost_in_employee_mappings(self):
...
+        })
+        self.assertFalse(self.project_global.analytic_account_balance)
+
+        self.env['project.task'].create({
+            'name': 'task A',
+            'project_id': self.project_global.id,
+            'planned_hours': 10,
+            'timesheet_ids': [
+                Command.create({
+                    'name': 'Mitchell Admin timesheet',
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_project.py:174]: It is not really the timesheet of Mitchell Admin. 😇 
You could just set `/` as `name` for this timesheet since it is not really important fro this test. 🙂 
```suggestion
                    'name': '/',
```

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I have made some comments. 🙂

File: addons/sale_timesheet/tests/test_project.py (hunk 10 above)
Comment (RaoufGhrissi) [reply addons/sale_timesheet/tests/test_project.py:156]: okay no problem

File: addons/sale_timesheet/tests/test_project.py (hunk 9 above)
Comment (RaoufGhrissi) [reply addons/sale_timesheet/tests/test_project.py:163]: yes i think my pr for this fix was passed! so it will work

**Review by xavierbol — COMMENTED** [review]
LGTM, just the commit message to change after that, it will look good to me. 🙂

**Review by xavierbol — APPROVED** [review]

File: addons/sale_timesheet/tests/test_project.py [hunk 12]
```diff
@@ -129,3 +129,50 @@ def test_compute_cost_in_employee_mappings(self):
...
+            'name': self.product_delivery_timesheet1.name,
+            'product_id': self.product_delivery_timesheet1.id,
+            'product_uom_qty': 1,
+            'product_uom': self.product_delivery_timesheet1.uom_id.id,
+            'price_unit': self.product_delivery_timesheet1.list_price,
+            'order_id': sale_order.id,
+        })
+
+        unit_amount = 6
+        expected_analytic_account_balance = self.employee_user.timesheet_cost * unit_amount * -1
```
Comment (IT-Ideas) [inline addons/sale_timesheet/tests/test_project.py:153]: I would find this more readable.
```suggestion
        expected_analytic_account_balance = -self.employee_user.timesheet_cost * unit_amount
```

**Review by IT-Ideas — COMMENTED** [review]
Thanks for your work 🙂

**Review by IT-Ideas — APPROVED** [review]
Thanks 🙂 
@robodoo r+
